### DISCUSSION TITLE
How does react-query handle overlapping requests to update cache

### DISCUSSION THREAD

**Discussion by caraclarke** [TanStack/query#4865] [Q&A] [answered]
**TL;DR:** If a useQuery response and pusher-js try to add the same piece of data to the cache at the same time, how does react-query handle simultaneous requests like this? A queue? Is there a chance the data could be added twice and be duplicated in the cache?

-- 

I am using react-query with with pusher-js library to get live updates as things change in my application to avoid the user having to refresh to see changes. I want to add a basic check to the pusher mapper to say "if this unique id is already present in the cache do not try to update" but I was wondering about a potential edge case and wanted to reach out here to see if this was a realistic concern or if this was something react-query handles behind the scenes.

My question is: 
If I have a response come in with data from a `useQuery` call and a push message come in with the same data, how would react-query handle those updates to the cache? Does it queue updates so one response would get written to the cache first and then the other would either overwrite it or be dismissed since theres no difference between the two pieces of data? Or is there any possibility that somehow both pieces of data could be added to the cache and we'd have duplicate entries? Since the keys are unique ids I wasn't too concerned about that happening but its been a debate on my team so wanted to throw it out there

**Comment by louis-young** [comment, marked as answer]
Hi 👋 

As far as I know, the query cache is essentially structured as a key/value map, with the key being the hash of the query key. I _think_ TanStack Query has de-duplication logic internally to deal with concurrent queries, but I don’t think this affects programmatic updates to the cache.

To attempt to answer the question, I imagine the query resolving successfully _after_ you’ve programmatically updated the entry in the cache for that query would result in the result of the query being kept in the cache, and the data you’ve programmatically set being kept in the cache if this happens _after_ the query has resolved successfully. I don’t know what’d happen if these two things happened at the _exact_ same moment in time. You could also consider canceling the query before performing the programmatic update if that's the behaviour you're looking for to avoid conflict. 

Take this with a pinch of salt as I’m not a maintainer; it’s what I _think_ would happen based on what I’ve seen and I thought it might help.

This is something you could probably test fairly accurately yourself if you were interested in doing so.

I’ll leave this open in hope of a more authoritative answer from someone with better knowledge of the internals but hopefully this sheds some light 🙂

**Reply by TkDodo** [reply]
>  I don’t know what’d happen if these two things happened at the exact same moment in time.

They cannot happen at the exact same time because JavaScript is single-threaded. One always comes later, and that one wins :)

**Reply by louis-young** [reply]
Of course, silly me... Thanks Dominik!
